dora_node_hub/dora_diff_drive_controller/src/odometry.rs

```diff
@@ -0,0 +1,158 @@
+                x: 0.0,
+                y: 0.0,
+                heading: 0.0,
+            },
+            linear: 0.0,
+            angular: 0.0,
+            previous_time: SystemTime::now(),
+            previous_left_wheel_position: 0.0,
+            previous_right_wheel_position: 0.0,
+        }
+    }
+
+    pub fn update(&mut self, left_wheel_position: f64, right_wheel_position: f64) {
+        let now = SystemTime::now();
+        let elapsed = now.duration_since(self.previous_time).unwrap_or(Duration::from_secs(0));
+        let dt_millis = elapsed.as_millis();
+
+        if dt_millis < 100 { // todo, this should 10!!!!!!!!!!!!!!!!!!!!!
+            return; // Ignore updates that are too close together
+        }
+
+        // Calculate the change in position of each wheel
+        let left_wheel_diff = left_wheel_position - self.previous_left_wheel_position;
+        let right_wheel_diff = right_wheel_position - self.previous_right_wheel_position;
+        print!("#########################################################################################3\n");
+        print!(
+            "Odometry update: left_wheel_position: {:.3}, right_wheel_position: {:.3}, dt_millis: {}\n",
+            left_wheel_position, right_wheel_position, dt_millis
+        );
+
+        print!(
+            "Odometry update: left_wheel_diff: {:.3}, right_wheel_diff: {:.3}\n",
+            left_wheel_diff, right_wheel_diff
+        );
+        let dt_secs  = dt_millis as f64 / 1000.0;
+
+        print!("dt_secs: {:.3}\n", dt_secs);
+        print!("self.wheel_radius: {:.4}, self.wheel_separation: {:.3}\n", self.wheel_radius, self.wheel_separation);
+        // Obtain velocity. 
+        // Note that there is no division by dt as it would be canceled out by the multiplication by dt when updating the pose.
+        self.linear = (left_wheel_diff + right_wheel_diff) * self.wheel_radius / 2.0;
+        self.angular = (right_wheel_diff - left_wheel_diff) * self.wheel_radius / self.wheel_separation;
+
+        print!(
+            "Odometry update: linear: {:.3}, angular: {:.3}\n",
+            self.linear, self.angular
+        );
+
+        let direction: f64 = self.current_pose.heading + self.angular; // * 0.5;
+
+        print!("direction: {:.3}, curren_pose.heading: {},  \n", direction, self.current_pose.heading);
+        print!("cos(direction): {:.3}, sin(direction): {:.3}\n", direction.cos(), direction.sin());
+        // Update the pose based on the velocities
+        print!("Before update: Pose2D(x: {:.3}, y: {:.3}, heading: {:.3})\n",
+            self.current_pose.x, self.current_pose.y, self.current_pose.heading);
+
+        self.current_pose.x += self.linear * direction.cos();
+        self.current_pose.y += self.linear * direction.sin();
+        self.current_pose.heading += self.angular;
+        // Normalize heading to the range [-pi, pi]
+        // self.current_pose.heading = normalize_angle(self.current_pose.heading);
```

Comment:
nit: f64 implements `atan2` which you can use to normalize: See https://doc.rust-lang.org/nightly/std/primitive.f64.html#method.atan2

Reply:
Hey! I understand atan2 returns a normalized angle. But in that case, the input has to be the cathetus of a right-angled triangle. In this case, we need to normalize the angles because we are performing +/- operations. How does atan2 apply to that?

Reply:
You have a point.
FTR: Continuation at https://github.com/Ekumen-OS/andino-rs/pull/54#discussion_r2273440453